Amazon S3 backend support

qris opened the pull request:
I will add patches from my work-in-progress branch to this pull request, to allow Travis to test them before merging.

qris committed:
Remove duplicate QDBM logging macros from BackupClientInodeToIDMap.cpp

qris committed:
Make IOStream::CopyStreamTo honour timeout

Throw an exception on network timeout.  This is safer than returning false,
which isn't checked in many places, and allows us to return the number of bytes
copied instead.

(cherry picked from commit e38d3d771ebb303e75e635e618f4d1e41370aa23)

cmorty commented:
Travis should also build every push to your s3_support_merge, unless you have configured it otherwise.

qris committed:
Remove unused intercept code from testbbackupd.

(cherry picked from commit 56631b8da4f8f22f291323d640560aca43f32b29)

qris committed:
Move S3BackupFileSystem class into its own file.

Create BackupFileSystem, an abstract interface for reading and writing
files and directories, abstracting over RaidFile, S3, FTP etc. as a parent
class for S3BackupFileSystem.

qris committed:
Enable compiler error on classes with non-virtual destructors.

I found a serious problem where subclasses didn't clean up their resources
properly when destroyed, if they inherited from a parent class with a
non-virtual destructor. This caused BackupFileSystem's NamedLock not to be
released. This commit:

* fixes the error where it occurred;
* enables static checks for this error when compiled with GCC;
* promotes the resulting warnings to errors, to ensure that this problem
  cannot reoccur.

qris committed:
Improve exception handling in S3Simulator, S3Client and HTTPServer.

Centralise error response generation in S3Simulator, driven by throwing and
catching exceptions.

Improve error messages returned by HTTPServer.

Improve exception messages thrown by S3Client.

Add a method to BoxException to quickly check what type of exception it is,
and a macro to make this even shorter.

Show the exception type as well as subtype in GetMessage().

qris committed:
Improve exception messages when configuration keys are missing or invalid.

Include a message in the exception, and reuse code to make them consistent.

Whitespace cleanup.

(cherry picked from commit 8589bd9a2d977a86be2917a382c2559858e63bff)

qris committed:
Fix negative estimated remaining time due to 64-bit integer overflow on large files.

(cherry picked from commit b27d4e7c61b04db054399435b5c57df1f9b1832b)

qris committed:
Use BackupFileSystem locking in testbackupstorepatch

Take BackupFileSystem locks in testbackupstorepatch to hopefully prevent race
conditions leading to store corruption and unreliable tests when using
housekeeping out-of-process.

(cherry picked from commit 416dcacb332297ac9a832cdad6bfa1e6739a52f7)

qris committed:
Remove link and unlink macros on Windows.

They provided source-code compatibility with the POSIX versions of these
functions, but they conflict with Boost. Replace with EMU_LINK and
EMU_UNLINK everywhere.

qris committed:
Travis build script: add extra debugging.

This script can be useful as a shortcut for command-line builds, but not if you
forget to set TEST_TARGET, so now it errors if you do.

qris committed:
Fix overloaded virtual function warnings.

Revealed by enabling the -Woverloaded-virtual compiler option.

qris committed:
Whitespace cleanup

(cherry picked from commit 88f7f4ef025a32680d0545571e0e5bdf9e0d5c3c)

qris committed:
Refactor test/s3store to make check_new_account_info() reusable.

(cherry picked from commit 2f597e11ccf56129319851f767581de1055da53e)

qris committed:
Fix warning about snprintf() being deprecated on Windows.

(cherry picked from commit eeb2a127fe1410a871f3943651fa363edff943d0)

qris committed:
Refactor common HTTP header handling

Move the common code out of HTTPRequest and HTTPResponse into the new
HTTPHeaders class, a container for headers with some sensible HTTP-specific
behaviour.

(cherry picked from commit 546c0d068934e8fc5ba53f32bdf61b3a6841c65b)

qris committed:
Fix incorrect timeout calculation in GetLine

Attaching a debugger to a process waiting in GetLine would previously cause the
read() to return immediately with EINTR, which we never retried. Now it retries
until the original timeout has expired.

(cherry picked from commit e997d092ad75aac98b5814c1b24d3ac0f8ae8e4a)

qris committed:
Add tests for S3Simulator and S3Client

qris committed:
Add MD5DigestStream and test

(cherry picked from commit 3f0538bf1a2115c399f09a1735e7e9f3b426737e)

qris committed:
Refactor S3Simulator to better support merge of SimpleDB support

(cherry picked from commit 35b581b3ac54769f5b343c995435a7c7b15041e7)

qris committed:
Move BackupStoreObjectDump.cpp to lib/backupstore

It doesn't depend on anything in lib/backupclient, so there's no need for it to
be there.

(cherry picked from commit 85fda42a5164cc574b9a9b0335d0cb14bbb802b0)

qris committed:
Add Boost, needed for XML support in S3Simulator and S3Client

Boost will be needed on all platforms for S3 support at least. It's currently
mandatory, but could be made optional if necessary. XML support of some kind is
required both for S3's ListBucket command (which returns XML) and for SimpleDB
support in the S3Simulator, since the SimpleDB API uses XML.

I considered alternative XML libraries to Boost:

* https://stackoverflow.com/questions/170686/what-is-the-best-open-xml-parser-for-c
* https://stackoverflow.com/questions/9387610/what-xml-parser-should-i-use-in-c

The most commonly used one (libxml2) is huge, ugly, slow and hard to use, and
we don't need any of the extra features.  TinyXML would meet our needs, but
Boost is more likely to be available on more platforms, is well respected, and
brings other (unrelated) features that we might want to use in future.

(cherry picked from commit 1a9deec05151281e840b7df5a1244453c0111042)

qris committed:
Add an assertion to CollectInBufferStream

Ensure that we don't try to write a negative number of bytes.

(cherry picked from commit 755242a46275384592281878ad6cce8dc7883be7)

qris committed:
Remove unwanted prototype for HeadObject()

(cherry picked from commit 71d3d970b00ee85408e8ea5166fe46fdbac82ce1)

qris committed:
Allow control of network timeout in HTTPResponse::Send()

(cherry picked from commit aabb13634e2ef79eb2843f8ab2f0cbd7be267c38)

qris committed:
Log every request made using S3Client (for now)

(cherry picked from commit 699a70598942c1422f568843176542b9b302c1dd)

qris committed:
Move BackupStoreAccountControl code to BackupAccountControl file

(cherry picked from commit 2891416b675cdc346713a26c6e357fba302bbc89)

qris committed:
Detect which debugger to use in t-gdb scripts: lldb or gdb

Should make debugging on Mac and BSD less painful.

Make t-gdb fail with an error when no debugger was detected.

(cherry picked from commit da98a8183737fb1fcb658cebd9612791a4750c4c)
(cherry picked from commit ea6bc722131222d1fad08521e61f552f16941b01)

qris committed:
Remove *.vcxproj.user from disposable CMake build directory

(cherry picked from commit 9df71c89a0fc935ee9965ddd8c4bcc2e1ad769a4)

qris committed:
Eliminate conditional code for Windows file handles from NamedLock

Should improve code readability and robustness.

(cherry picked from commit e773e4ccefcdab8f01b1a3bf7d828c3bb39eaaa7)

qris committed:
Fix emulated readdir() d_type to return DT_FILE or DT_DIR like Unix does

We used to abuse dirent.d_type to store Windows attributes, but we can now just
add a new attribute to store these (win_attrs), so I have.

qris committed:
Create a job object during Windows tests to kill abandoned daemons

Should prevent some cases of tests hanging on AppVeyor due to child
process daemons that were started and not stopped.

Delete files between tests using much more similar code (and the same
list) on Windows and Linux, to reduce risk of divergence causing tests
to pass on one and fail on the other.

qris committed:
Don't bother with HAVE_DEFINE_PRAGMA, it's unnecessary and doesn't always work

(cherry picked from commit 83b31dac888f91843ec79eea028e3358d303aa92)

qris committed:
Enable memory testing under Darwin, as it does work now

(cherry picked from commit 86dc5ff3666c934dcf4fce261a93431b78bf018f)

qris committed:
Store and log full stack traces for leaked memory allocations

Don't dump stack backtrace when logging memory leaks to a file

Since the backtrace goes to the console instead, it's confusing to have them in
different places.

(cherry picked from commit 6e7df4dfdc06eaff2dd0ef51741df3cf6ba319ea)
(cherry picked from commit 270cd1f16fa8a32b07ae0b9762a417db7b95b82f)
(cherry picked from commit c627e9547fb8a191ea36780b140fd0516a0bffa8)

qris committed:
Fix build on SmartOS by switching back to GNU Make

I have no idea where I got the idea that there was a "bmake" command on SmartOS
that is BSD make, or why that would be better than gmake (GNU make). I can't
find any reference to the existence of such a thing. So I've just switched back
to using gmake unconditionally.

qris committed:
Hopefully fix test/raidfile on Solaris.

64-bit Solaris has _FILE_OFFSET_BITS defined (to 64), which was wrongly causing
the substitute lseek in lib/intercept to enter the 32-bit branch.

qris committed:
Remove unused code: BoxDebugTrace() and BoxDebug_printf()

qris committed:
Refactor winsock initialisation on Windows

Automatically initialise winsock in all binaries. Should fix issues with
bbstoreaccounts not working on Windows.

qris committed:
Fix exit status from tests: could previously return 0 even after failures

qris committed:
Reduce precision of common timer test

Ideally timers would be perfectly accurate and we could sleep for 1.0 seconds,
but on OSX in particular they could fire 50-100 ms late (I've seen 4 ms in
practice) and we don't want the tests to fail because of this, because we don't
really need that kind of precision in practice. So we reduce the timer
intervals by 100ms to be safe.

(cherry picked from commit 8d02eebce553ed822e0fcd60d6e319384e15ba4b)

qris committed:
Allow 1 extra second in test_store_error_reporting

Hopefully this will solve the random test failures on Travis.

qris committed:
Add support for M4 CXX flag detection in CMakeLists

Enables all flags that are detected by M4 builds to be detected by CMake builds
as well (except for -Wall which enables a bunch of junk).

(cherry picked from commit 2ea8a9706d79a386729c7b2d814089e90bac6926)

qris committed:
Fix delete of incomplete type in BackupAccountControl.h

This has been flagged up by MSVC for some time. I've been unable to make GCC or
Clang do so.

qris committed:
Log time taken by each test case in a test suite.

Should help diagnose why test/backupstore sometimes takes >500 seconds on
Travis, when it normally takes ~120 seconds.

qris committed:
Add support for RelWithDebInfo builds using CMake

Debugging release tests is pretty difficult without debug info. This
release-compatible build makes it easier.

qris committed:
CMake: make readline support configurable, not automatic

Otherwise if it's detected on the system (e.g. if you install GnuWin)
then it would always be enabled, but linking against GnuWin headers
breaks the build.

qris committed:
Win32 build: upgrade to _WIN32_WINNT 0x600 (WinXP and above)

This should enable us to use Windows dbghelp functions to get stack
traces on exceptions, including demangled names.

This required replacing our custom "struct pollfd" with
EMU_STRUCT_POLLFD everywhere, to fix the conflict with newer versions of
winsock2.h.

qris committed:
Win32: add experimental stack backtraces with symbol lookup

(cherry picked from commit f4685e72c2c7e274533c1cb149180809631a1da7)

qris committed:
Make generated files depend on their generators under CMake

If makeprotocol.pl.in or makeexception.pl changes, then we should rebuild
the source files that are built by them (and therefore anything that depends
on those source files, etc).

(cherry picked from commit f9818932b31120f139500cc524563fe8483b072d)
(cherry picked from commit 53d1a6c05ed4c39cb2eef34efb52a544ec4e02a1)

qris committed:
Modify command-line -L option to override log level for a specific file

This allows enabling TRACE logging for a specific file (e.g. to enable
exception backtraces for that file) without enabling TRACE globally.

Replace HideFileGuard with LogLevelOverrideByFileGuard.

(cherry picked from commit 29f74be06351b6ce9a872765db63171ad902b596)

qris committed:
Use filename where exception was thrown in DumpStackBacktrace

Allows us to turn on exceptions for specific files, rather that
filtering them all as Utils.cpp.

(cherry picked from commit 3590e9745dac2f25267bd755ed3d554423395330)

qris committed:
Enable backtraces and timestamps in all tests when building with CMake

(cherry picked from commit 166e39c11d242ebed6b62fc3f126af48345a22ca)

qris committed:
Use the correct preprocessor define for release builds

(cherry picked from commit 085149bf1c2c287ee744f2e094a99a77fc89dc60)

qris committed:
Show IP address as well as hostname for ServerException::SocketOpenError

(cherry picked from commit 1a51adcd6be000f09fa280e42fb12e5631cd69b9)

qris committed:
bbstoreaccounts: add support for standard command-line logging options

Tidy up usage text, and fix erroneous memory leak reported after calling
PrintUsageAndExit().

(cherry picked from commit 1a209b5c3c9cdd0ed3a92d2c884706a580b13844)

qris committed:
Update banner text, and shorten compiled-in Git version numbers

Shorten compiled-in version numbers based on Git revisions.  Git revisions are
often quoted as 7 hexadecimal digits, so let's do that here too.

Add sanity to banner texts of command-line utilities.

(cherry picked from commit 9afc3ca99d9cddd6f9a01fd5c0baa63ee16a449a)
(cherry picked from commit c10913a35c155fb1efbc2c32ba0db45c580c4fad)
(cherry picked from commit 0ce96b4c356b9f94b46e0c2842fdc81704c77f76)

qris committed:
Perl buildsystem: add test prepare targets

Similar to the CMake prepare targets, these are useful when intending to
debug tests or run a server using a standard test configuration, which
requires the testfiles directory to be copied into the binary output tree.

(cherry picked from commit 2afecdd57a3bcf42c116c8535843039a1596feac)

qris committed:
Fix compile warnings about printf format strings in DebugMemLeakFinder

qris committed:
Improve NamedLock code readability, reliability and logging

Define our own preprocessor defines for lock methods, test that we cover every
case at compile time ad that they have the semantics that we expect (reentrant
or not).

Add support for F_OFD_SETLK, a non-reentrant lock type offered by Linux, and
ensure that this and FLOCK are chosen in preference to F_SETLK locks, which are
reentrant.

(cherry picked from commit de1911b132310a1d5f1073928036db00abe0ef20)
(cherry picked from commit 96136adf5dbe38ef92ad3f9490b549c11dd91fd4)

qris committed:
Lock store while corrupting it in test/backupstorefix

Should hopefully reduce risk of random test failures/race conditions.

(cherry picked from commit 0f31359ca8ebbd6ddfd5f2d042c8c23af4d67997)

qris committed:
Normalize all the line endings

qris committed:
Tell Git to treat .cpp files as text too.

qris committed:
Tell Git to ignore some more generated files

qris committed:
Add Boost to Travis build environment, and OSX builds

qris committed:
Update versions of dependencies for Windows unibuild, add Boost

* Add a variable in appveyor.yml for the CMake out-of-source build tree.
* Update the directory name in commands in appveyor.yml to match new CMake unibuild tree structure.
* Increase build verbosity slightly.
* Disable GUI crash dialogs that cause test hangs.
* Disable the AppVeyor cache that we don't use.

qris committed:
Install executables in test tree after building, not during "make install"

qris committed:
Relax timings on test_changing_client_store_marker_pauses_daemon

Also increase test verbosity to help debug the test if it fails again.

qris committed:
Improve error messages when PCRE detection fails (or is abused)

(cherry picked from commit 81528092fd51b6f81a21899bc205c369597c481d)

qris committed:
Add a platform-independent implementation of box_strtoui64

Replaces strtoull(), which doesn't exist on older Windows compilers (MSVC <
12), and in any case doesn't necessarily return a uint64_t, so this is a better
interface.

qris committed:
Fix confusion about URIs and paths in S3 filesystem

* Make it clear (with parameter names and comments) which functions take paths
  (relative to the configured base path), which functions take URIs (complete
  paths) and which functions take URLs (hostname, port and URI).

* Put all objects (files and directories) in the same "directory" tree on S3,
  so that we can list them all together and easily identify the type of each
  object (from its extension). Previously they were separated into "directories"
  (delimited prefixes) called "files" and "dirs" (unlike BackupStore
  filesystems).

* Reinstate hierarchical "directories" on S3, similar to BackupStore
  filesystems, since the reasons for doing this in RaidFile may well apply to S3
  and other compatible stores too. However the objects are still named with
  their complete object ID (unlike RaidFile) and a .file or .dir extension.

(cherry picked from commit 3dd21a3028cacea2ecf6ee3de2f32db3c5138589)

qris committed:
Fix quoting of CXXFLAGS and LDFLAGS in Makefiles

Previously, it was impossible to include an @ sign in these flags (for example,
to link to support --with-ssl-headers=/usr/local/opt/openssl@1.1/include/)
because it was interpreted as a list variable name by Perl. Now the variables
are evaluated single-quoted before being substituted into the Makefile output.

qris committed:
Replace RAND_pseudo_bytes with RAND_bytes

RAND_pseudo_bytes is less secure and is now deprecated.

qris committed:
Add support for OpenSSL 1.1 and replace deprecated function calls

See https://github.com/boxbackup/boxbackup/issues/16 for details. Thanks to
Chris West @FauxFaux for the initial patch!

Use OpenSSL 1.1.0g for Windows superbuild.

Fix Windows build by removing calls to obsolete do_ms.bat
<https://stackoverflow.com/questions/39076244/why-there-is-no-ms-do-ms-bat-after-perl-configure-vc-win64a/39247560#39247560>.

Workaround for incorrect library suffixes searched by FindOpenSSL
<https://gitlab.kitware.com/cmake/cmake/issues/17604>.

Link OpenSSL statically and fix missing dependency on crypt32.lib.

(cherry picked from commit edd3687f067c68b131822e0064cdeff5bf7a3835)
(cherry picked from commit b003e009d1fccadf55a0f59a01c66f7b80f9b3e5)
(cherry picked from commit cc6e20466b857f7e20d816a2c20cb36f112398ab)
(cherry picked from commit 303c6405e3bc1e673dbc691792e0633742c01f4e)
(cherry picked from commit 4e24006f75142a254f1a7db76cf9da1e752d9283)
(cherry picked from commit 448ac48a89538f67f508f8d21931c5e6e90472c0)
(cherry picked from commit 00b3bb0697219ef8c742698dd43d53b538e362b1)
(cherry picked from commit 85e7efc3fa0477f60318d2cd2144503a9ea8feb9)

qris committed:
Declare Base64 decoding array as signed char

Need to use "signed char" explicitly to work around char being unsigned by
default on ARM, causing compile errors:
https://stackoverflow.com/a/31635045/648162

qris committed:
Make narrowing conversions an error

E.g. on ARM64 with unsigned chars, this is a real error, and causes the build
to fail:
https://buildd.debian.org/status/fetch.php?pkg=boxbackup&arch=arm64&ver=0.11.1~r2837-4&stamp=1485057548&raw=0

(cherry picked from commit 3433291fc5bfbbdc87385796c26c44699edf5c0c)

qris committed:
Make narrowing conversions an error

E.g. on ARM64 with unsigned chars, this is a real error, and causes the build
to fail:
https://buildd.debian.org/status/fetch.php?pkg=boxbackup&arch=arm64&ver=0.11.1~r2837-4&stamp=1485057548&raw=0

qris committed:
Stop Git from normalising line endings

Although recommended, "text" mode causes constant and unfixable conficts
that I give up fighting with.

(cherry picked from commit 25e445a6848a870aabd89a3b8a013265aa3cf17b)

qris committed:
Increase logging level and add timestamps in testcommon

Should help to debug frequent test failures such as Condition
[t2.HasExpired()] on OSX hosts on Travis.

(cherry picked from commit 61a8e7514e01bc05ea3ea77d110c16bff37a50f4)

qris committed:
Fix incorrect detection of the need to intercept lxstat

For some reason, #if defined __GNUC__ && __GNUC__ >= 2 was true on OSX,
resulting in intercepts not working properly.  It was also broken on MIPS64,
causing test/raidfile to fail with errors like: TEST FAILURE: Condition
[TestGetFileSize(stripe1fn) == fs1] failed at testraidfile.cpp:415

https://buildd.debian.org/status/fetch.php?pkg=boxbackup&arch=mips64el&ver=0.11.1~r2837-4&stamp=1485060478&raw=0

Now we detect when to intercept Linux' lxstat syscall using autoconf macros.
We also detect whether to intercept stat64 (instead of stat) based on whether
the stat64 function exists.

(cherry picked from commit a042c9951fabc1fb4f57e6d13d739192dda76025)

Fix d2775d5b

qris committed:
Fix check for a working struct dirent.d_type

Which I broke in 2007 and never noticed was broken until now. Oops!

qris committed:
Don't add CXXFLAGS to CFLAGS

Some CXXFLAGS such as enabling C++11 don't make sense for CC and will be
rejected, causing the compile to fail. We don't really test for flags being
compatible with the C compiler anyway, and we don't have enough C files to
care that much.

(cherry picked from commit 2ad9ee9012d4a7354c33ff0272c0768b952a9064)

qris committed:
Enable and fix raidfile intercepts on Mac OSX

qris committed:
Add comments to explain the three S3Client constructors

qris committed:
Use a named enum type for DirReadType (RaidFileRead::ReadDirectoryContents)

Allows checking for valid values at compile time, and allows the debugger to
show the symbolic constant instead of just its numeric value.

qris committed:
Build Win32/64 binary packages automatically

Every AppVeyor build on the specified branches that passes will be packaged
into a ZIP file and installer using CPack and uploaded to GitHub Releases.

Generate version number entirely in AppVeyor config, and build it into the
binary, and use it in the names of the created binary packages, and compile it
into the binaries.

qris committed:
Whitespace fixes

qris committed:
Remove Xamarin to remove 500 lines of junk from build logs

See
http://help.appveyor.com/discussions/problems/4569-the-target-_convertpdbfiles-listed-in-a-beforetargets-attribute-at-c-does-not-exist-in-the-project-and-will-be-ignored for details.

qris committed:
Reduce AppVeyor build verbosity by passing extra args to sub-CMake

qris committed:
Reorder lines in appveyor.yml to match actual execution order

qris committed:
Disable DEBUG to reduce CMake build verbosity on AppVeyor

qris committed:
test/backupstorefix: tag bbstoreaccounts check output in tests

(cherry picked from commit c31d7143273f2d5257ec690e894442a8398aaf21)

qris committed:
Increase default verbosity in some timing tests

qris committed:
Rename BackupStoreAccountsControl to BackupStoreAccountControl

For consistency with the base class, BackupAccountControl, and grammatical
reasons.

qris committed:
Add missing implementations of RaidBackupFileSystem methods

These should have been added whenever BackupFileSystem.h was added, which was
4db4deed3448b9c2b9e885a20cb1a777632464d6, although it didn't really belong in
that commit either. But it's too late to change history now. Adding the missing
method definitions makes it possible to instantiate a RaidBackupFileSystem.

qris committed:
Refactor BackupStoreAccountControl to be for a specific account ID

This allows other BackupAccountControl implementations to implement the same
interface without needing to support account IDs, which are very
store-specific.

qris committed:
Remove debugging code to list files during test_multiple_uploads

Revert "Try listing files to help debug random failures of test_multiple_uploads()."

This reverts commit e302ded5b7d40fe88d65808564f4db9df2a6f878.

qris committed:
Custom build system: fix autogen dependencies

Remove useless inclusion of Makefile.extras with paths in subdirs that will
never match anything, and replace those inclusions with recursive Make calls to
build those targets. This allows changes to exception and protocol files to be
detected automatically, and the C++ code regenerated.

qris committed:
Make ObjectExists and BackupStoreCheck use object_exists_t enum

qris committed:
Add GetTempDirPath() for portability

On Unix, check TEMP, TMP and TMPDIR, and if no env var is set, fall back to
/tmp.  For some reason it's becoming increasingly common for Unixes not to set
TMP* to anything.

qris committed:
Fix include ordering in testcommon.cpp

qris committed:
Refactor testcommon to split into independent tests

Remove redundant NamedLock tests.

qris committed:
Add force parameter to RemovePrefix and RemoveSuffix

qris committed:
Remove BackupStoreContext::ReceivedFinishCommand

Replace with CleanUp(). We don't need two methods that do almost the same
thing.

qris committed:
Refactor BackupStoreContext::GetBackupStoreInfo

Almost all access to mapStoreInfo is now via GetBackupStoreInfo(Internal), in
preparation for the refactor that will move the main copy of the
BackupStoreInfo to the BackupFileSystem instance instead.

qris committed:
Enable C++0x support

We need C++0x features enabled for "auto" to work, and using Boost properly
seems to require this. Use "auto" in a few places to prove that it works.

qris committed:
Refactor to create BackupFileSystem::GetBackupStoreInfo

Extract common code from filesystem-specific implementations, and rename them
to GetBackupStoreInfoInternal().

qris committed:
Refactor BackupFileSystem::GetLock()

Call subclass implementations of TryGetLock(), and throw an exception if we
fail to get a lock instead of returning false.

qris committed:
Refactor test code into process_is_running()

We need the ability to check whether a process is still running to check for
stale S3 account locks in SimpleDB, so move this code out of Test.cpp and into
Utils.cpp, for use outside of tests.

qris committed:
Implement compare_lists and test_equal_maps in lib/common/Test.cpp

qris committed:
Add static method HTTPQueryDecoder::URLEncode

This will be needed for request parameter encoding and for SimpleDB.

qris committed:
Add query parameter editing and encoding to HTTPRequest

qris committed:
Add support for virtualhost names in S3Client

This is a precondition of support for both S3 and SimpleDB in the S3Simulator,
since otherwise we can't tell the requests apart.

qris committed:
Add simulator and client for SimpleDB

This will be used to support reliable instant locking of S3 accounts.

qris committed:
Add tests for S3Simulator and S3Client

qris committed:
Whitespace and comment cleanup

qris committed:
Add support for S3-compatible ETag headers to S3Simulator

qris committed:
Rename setup_test_bbackupd() in test/bbackupd

The name of this utility function was confusing with the SETUP_TEST_BBACKUPD
macro which served an entirely different purpose, so I renamed it to
prepare_test_with_client_daemon().

qris committed:
Add initial (non-functioning) keepalive support in S3Client

(cherry picked from commit 3c62a5be57e0a7b7100b44c41cc8c69c20c92fa9)

qris committed:
Add support for If-None-Match header to S3Client and S3Simulator

This allows the client to only download a file if it doesn't match the one in
the local cache, which makes it possible to use the cache very efficiently.

qris committed:
Make GetBlockSize() use BackupFileSystem

Refactor BackupAccountControl, BackupStoreContext and BackupCommands to call
BackupFileSystem to get the block size of the filesystem. It's a start.

qris committed:
Protocol generator: record last error message received

Store this as a string (as well as the error code) so that we can report it to
the user, for example in test failure messages.

qris committed:
Improve error message on hostname lookup failure

Throw exceptions with a message, including the OS error, which means retrieving
it properly from the winsock library on Windows.

qris committed:
Increase precision of timing on Windows.

Should hopefully reduce random failures of timing-dependent tests which
rely on subsecond precision for accuracy.

(cherry picked from commit c845b2e39fffeb560983a301d810616a6495469d)
(cherry picked from commit ce7e000d43a91dcd0d39ee3a505521e700875556)

qris committed:
Switch PCRE back to an official release

qris committed:
Remove BoxBackup install step from Windows unibuild

We don't need to install anything in C:\Program Files\BoxBackup, and often
don't have permission to do that anyway.

qris committed:
Upgrade to OpenSSL 1.0.2l

Run OpenSSL tests after building it for Windows Unibuild

qris committed:
Improve exception messages from RaidFileWrite

Use THROW_SYS_FILE_ERROR to include the system error message in the exception
message. Replace calls to BOX_LOG_SYS_ERROR where previously used.

qris committed:
Travis CI: exclude (don't build) tags created by AppVeyor

AppVeyor creates new tags for each release (binary build), which have the same
git commit that we already built, so building it again is a waste.

qris committed:
Add ResponseCodeString shortcut to HTTPResponse

Useful for debug logging and exception messages.

Add the actual HTTP response to the exception message in
S3Client::CheckResponse, to help with debugging.

qris committed:
Add option to include parameters in HTTPRequest::GetRequestURI

This is needed to support the ListBucket command.

qris committed:
Fix FileExists in Utils.cpp

Testing for st.st_mode & S_IFLNK doesn't work on OSX, because IFLNK and IFREG
share bits, so regular files have a nonzero result and appear to be links. It
seems safer and more readable to use the mode test macros, e.g. S_ISLNK and
S_ISDIR.

qris committed:
Add ListBucket command support to S3Simulator and S3Client

qris committed:
Add support for the HEAD and DELETE methods to the S3Simulator and S3Client.

qris committed:
Rename local variable

First part of refactoring HousekeepStoreAccount to use BackupFileSystem.

qris committed:
Fix conflict over unlocked refcount DB

If BackupStoreCheck is are read-only, then it should not open a "temporary"
database in the old sense, because that does actually change the store: the
temporary file would conflict with any other process which wants to do the same
thing at the same time (e.g. housekeeping), and if neither process locks the
store, they will break each other.

Instead, we now distinguish between "potential" refcount DBs (of which there
can only be one at a time, and which require a lock to create) and really
"temporary" ones which are created in a temporary directory, can never become
permanent via Commit(), do not require a lock, and are unlimited in number.
Commit() will rename them, but still in the temporary directory, so it has no
effect on the store.

This requires some changes to the BackupStoreRefCountDatabase::Create interface
to add new flags.

qris committed:
Refactor BackupStoreRefCountDatabase to an abstract interface

This makes it easier to cleanly subclass in BackupFileSystem, to change its
behaviour by hooking into its Commit() method.

qris committed:
Move management of BackupStoreRefCountDatabase into BackupFileSystem

This is needed to allow BackupFileSystem to commit the database to storage on
request, and to cache it, which is needed for S3 support.

qris committed:
Initial refactor of BackupAccountControl to use BackupFileSystem

Abstract away some account operations using BackupFileSystem so that they can
be implemented differently for S3 accounts. In particular, account opening and
NamedLock usage were targets for this refactor. BackupStoreInfo creation and
opening ended up being partially addressed as well. This significantly reduces
the diffs to the s3_support branch.

qris committed:
Enable compiler warning -Werror=overloaded-virtual, and fix

qris committed:
Refactor HousekeepStoreAccount to use BackupFileSystem

This involved moving big chunks of functionality into RaidBackupFileSystem, for
example CombineFile, CombineDiffs and RaidPutFileCompleteTransaction. Not much
is implemented for S3BackupFileSystem yet.

xkr47 commented:
Soooo what's up with this one? :)

qris commented:
@xkr47 very basic support is nearly done, I'm carefully splitting up and merging in the changes that I've made in order to keep history clean and help with debugging. There is a bit more work to do on merging patches (housekeeping), performance optimisation (local caching) and of course I need to test with the real

qris commented:
With the real S3. So far I've been working only with the simulator which I wrote to help with debugging and testing.

qris committed:
Refactor BackupStoreCheck to use BackupFileSystem, part 1

Pass the BackupFileSystem from BackupAccountControl to BackupStoreCheck, and
use it for locking the account (actually a no-op because it's already locked).
Trivial, but it's a start.

qris committed:
Refactor BackupStoreCheck to use BackupFileSystem, part 2

Use BackupFileSystem to access the refcount database.

qris committed:
Refactor BackupStoreCheck to use BackupFileSystem, part 3

Use BackupFileSystem to access the per-account BackupStoreInfo.

qris committed:
Refactor BackupStoreCheck to use BackupFileSystem, part 4

Use BackupFileSystem to EnsureObjectIsPermanent.

qris committed:
Refactor BackupStoreCheck to use BackupFileSystem, part 5

Use BackupFileSystem to Get and Put Files and Directories.

qris committed:
Refactor BackupStoreCheck to use BackupFileSystem, part 6

Use BackupFileSystem to delete corrupt objects.

qris committed:
Refactor BackupStoreCheck to use BackupFileSystem, part 7

Move most low-level filesystem structure check code into BackupFileSystem, and
add specialised implementation for S3.

qris committed:
Refactor BackupStoreCheck to use BackupFileSystem, part 8

Remove last vestiges of RaidFile from BackupStoreCheck.

qris committed:
Rename some local variables in BackupStoreCheck

qris committed:
Add configuration for S3 local cache directory

qris committed:
Fix repeated calls to TryGetLock() when already holding lock

qris committed:
Implement RefCountDatabase methods in S3BackupFileSystem

qris committed:
Fix writeback of refcount databases on BackupFileSystem destruction

Call ReleaseLock in the child destructor (not the parent) to allow the refcount
database to clean itself up (normally involving calling
BackupFileSystem::SaveRefCountDatabase) before the BackupFileSystem is too
destroyed to be able to fulfil the request.

qris committed:
Wait for housekeeping to finish before bbstored exits

Should fix random test failures due to housekeeping process realising that the
master has asked it to terminate, and logging that it has, after the master has
already died and the test finished, confusing the test runner with its extra
output. Should also help make the occasionally-reported housekeeping crashes
more obvious and easier to debug.

qris committed:
Wait for housekeeping to finish before bbstored exits

Should fix random test failures due to housekeeping process realising that the
master has asked it to terminate, and logging that it has, after the master has
already died and the test finished, confusing the test runner with its extra
output. Should also help make the occasionally-reported housekeeping crashes
more obvious and easier to debug.

(cherry picked from commit 0100d3185abf10b17ff7727f2c0f2d8fb6b724d4)
(cherry picked from commit 2cc729b45d60f3c9b920a459538171882a564b0d)

qris committed:
Remove useless Get/Put/HeadObject methods from S3BackupFileSystem

These should not have been part of the API, and are not needed since mrClient
has the same methods.

qris committed:
Refactor BackupAccountControl to use BackupFileSystem

qris committed:
Refactor BackupStoreFile::VerifyStream from write to read side

This allows us to wrap an existing stream (not necessarily a file, e.g. a
streaming HTTPRequest) in a VerifyStream and pass that stream to a
S3PutFileCompleteTransaction, which isn't just going to CopyStreamTo()
somewhere, but do something more complex to it.

qris committed:
Fix a long-standing bug in the set_refcount() test helper function

Previously it would truncate the expected refcount list at the first zero entry,
instead of just removing zero entries from the end.

Also move some test code to test_server_commands which should have been there,
since they were causing issues with the check for the refcount database in
test_server_housekeeping. It's hard to disentangle these two changes.

qris committed:
Add extra refcount checks to test_server_housekeeping

qris committed:
Refactor BackupStoreContext to use BackupFileSystem

qris committed:
Fix Makefile dependency tracking of configured files

This allows protocol files to be regenerated automatically if
makeprotocol.pl.in changes, and similarly exception files and
makeexception.pl.in.

qris committed:
Enable and fix warnings in infrastructure/makebuildenv.pl

qris committed:
Refactor GetLine

The class name was a problem, because the method that we want to call on it is
also called GetLine, which looks too much like a constructor in C++, so I
renamed it to LineBuffer.  I haven't yet renamed the child classes.

Also, it was impossible to distinguish between the different causes of
GetLine() returning false (timeout and signal) and EOF could only be checked by
calling a different function. Now all of these causes throw different
exceptions.

This really fixes the problem with attaching a debugger to a process waiting in
GetLine, which would previously cause the read() to return immediately with
EINTR, which we never retried because it looked like a timeout.

qris committed:
Implement placement operator delete

"If the object is being created as part of a new expression, and an exception
is thrown, the object’s memory is deallocated by calling the appropriate
deallocation function. If the object is being created with a placement new
operator, the corresponding placement delete operator is called—that is, the
delete function that takes the same additional parameters as the placement new
operator. If no matching placement delete is found, no deallocation takes
place."

So to avoid memory leaks, we need to implement placement delete operators that
correspond to our placement new, which we use for leak detection (ironically)
in debug builds.

(cherry picked from commit 40cf96207139a31466b98a93298cef4f436eea6b)
(cherry picked from commit 242fac41c63309f274992825cd0baef9046e1110)

qris committed:
Enable support for S3 virtual hosts in BackupAccountControl

This enables the S3 bucket name to be included in the virtual hostname, for
example bucketname.s3.amazonaws.com, when connecting to a different hostname
such as localhost. This makes bucket names usable in tests with
BackupAccountControl.

qris committed:
Properly implement S3BackupFileSystem::GetPermanentRefCountDatabase

qris committed:
Remove always-false assertion in S3BackupFileSystem::GetDirectory

This was designed to stop untested code from running. Now we are ready to use
this code, with tests soon to come.

qris committed:
Make BackupStoreContext::ClearDirectoryCache() public

Needed to clear the cache in backupstore tests specialised for S3.

qris committed:
Add store test helpers for specialised tests

qris committed:
Test that 100-continue headers are sent and received correctly

This tests that HTTPRequest::SendHeaders includes the "Expect: 100-continue"
header only when expected. A separate test calls HTTPRequest::SendWithStream
which sends these headers, and therefore checks that both sides correctly
handle such exchanges.

qris committed:
Test that the S3 backend properly implements basic commands

This extends test_server_commands() to the S3 backend implementation, using
test specialisation.

qris committed:
Fix rare crash while tearing down S3BackupFileSystem

ReleaseLock() could fail if it can't contact the S3/SimpleDB server, but
destructors aren't allowed to throw exceptions, so catch this case and log
a warning instead (because we failed to release the lock that we were
holding).

(cherry picked from commit 6226c014dd2988a6a29cd21f6cc29fa0e7816607)

qris committed:
Improve exception handling in single-process servers

Don't allow an uncaught exception to kill the whole server. This makes it
consistent with multi-process servers, where death of a child worker does
not stop the main server process, so it continues to serve requests.

This makes the HTTPServer robust to exceptions (such as the child
disconnecting while the server is streaming a response) on Windows.

(cherry picked from commit 4a515b9ce8e616b357dc07bee27aa7b3692c3602)

qris committed:
Windows syslog() emulation: improve error messages

If syslog can't log to the Windows event log for some reason, instead of
just printing the log message to the console, include some explanatory
text about what went wrong.

(cherry picked from commit c492c90062d2e88db1016f13bd912629477a8a46)

qris committed:
Completely disable syslog logging in tests

(cherry picked from commit 3a8ca0c97e7346912889df72a1a87b80b430fe7a)

qris committed:
Enable overriding log level by category as well as filename

This extends the -L command-line option to allow overriding by
filename/category=level as well as the existing support for overriding by
filename=level.

(cherry picked from commit 4421c53ab59adb494308f9c9f881a6a57bfd571c)

qris committed:
SocketStream: use THROW_SOCKET_ERROR for better messages on Windows

(cherry picked from commit feddbef8052208508a2e542a1a70c23138f476b8)

qris committed:
Increase timeouts in S3Client to help with debugging

(cherry picked from commit cc14b3b4b9ec509233000c7d0a995580afd97f15)

qris committed:
S3Simulator: add some TRACE logging to help debug 100-continue responses

(cherry picked from commit ca39b88d4a73b64de68400dfa29270f203fd0f14)

qris committed:
ServerStream and test/backupstore: improve comments slightly

(cherry picked from commit c50aac67ee6ce6af9d63d5b96290585f2c5b7c21)

qris committed:
Clear up Daemon command-line arguments -D and -F

By separating their roles (which used to overlap) it's now possible to run a
forking server in the foreground, or a non-forking daemon in the background,
which can be useful for debugging (for example, to more closely simulate how
test daemons work on Windows when running on Unix). -DF is now needed to
have a non-forking server in the foreground.

(cherry picked from commit b31c0e4abd4875bb25ff7a918ed5bc7f7faf595d)

qris committed:
Fix SimpleDB timeouts: should be in milliseconds, not seconds

(cherry picked from commit 24bc49f30b9ce54d5024777f9c34d70a476d9663)

qris committed:
Add support for TcpNice on OSX, remove on Windows

I'm not convinced that it does anything on Windows without an RTT estimate,
which we can't get without administrator access:
https://msdn.microsoft.com/en-us/library/windows/desktop/bb485738(v=vs.85).aspx

(cherry picked from commit b646d22f415bbc8f5b5b065ed7462166591404f2)

qris committed:
Move functions around to reduce diffs to s3_support

qris committed:
Fix compile error in BSD getopt_long.cpp

error: '&&' within '||' [-Werror,-Wlogical-op-parentheses]

qris committed:
Generalise BackupFileSystem::GetLock and reuse it

This reduces code duplication (multiple implementations of locking retries) and
also enables centralised logging of locking attempts, which can be more easily
fine-tuned.

(cherry picked from commit 0469e4a3bd7318547294ec5309aabffa488fc67f)

qris committed:
Allow using LogLevelOverrideByFileGuard as a guard

After calling its Install() method, the guard will be installed in the central
logging configuration, and will automatically be uninstalled when the object
goes out of scope and is destroyed.

(cherry picked from commit 2b5b65b2dccfa2a88688e0103dc4996a039c515c)

qris committed:
Add detailed logging of BackupFileSystem locks

This logging is categorised, so it can be enabled with fine-tuning using the -L
command-line option.

(cherry picked from commit 7a2572c446b4a95f18eda04e72f6aabfd9544e04)

qris committed:
Enable detailed logging of locking in test/backupstorefix

This should help diagnose occasional locking problems causing test failures on
AppVeyor.

(cherry picked from commit 94f75ac0476c2264ed80f51c5d4014639dccb407)

qris committed:
Suppress exception messages in LineBuffer::GetLine

These exceptions are used to communicate state to the caller, and should not be
treated as errors in common cases.

(cherry picked from commit 86f3e9c0f4916741bbed20e04a4b4f2d71bb4f32)

qris committed:
Fix cleanup of backupstore specialised tests

In case of exception, ensure that teardown code is always run, to release
filesystem lock before deleting files in standard tearDown().

(cherry picked from commit 69f205bafe925548d799df13d01d2e8ea5e5e7bc)

qris committed:
Increase time allowed for backup in test_store_error_reporting

Should increase reliability of tests running on Travis.

qris committed:
Enable precedence warnings as errors (-Werror=parentheses)

qris committed:
Remove invalid use of null references (undefined behaviour)

Enable the relevant compiler warning as an error, if supported
(-Werror=undefined-bool-conversion).

http://www.gotw.ca/conv/002.htm
http://stackoverflow.com/questions/2165078/a-reference-can-not-be-null-or-it-can-be-null

(cherry picked from commit f2911acac0c8375a08ecc0a55f853a5a59c4d511)

qris committed:
Enable errors on unused private fields (-Werror=unused-private-field)

Fix compile failure due to unused private field.  In CMake release builds,
RaidBackupFileSystem::mDiscSet was initialized but never read, and this was
detected by -Werror,-Wunused-private-field.

Now the field is totally compiled out of release builds, and by using the
correct preprocessor define, CMake and the Perl buildsystem now behave the
same.

(cherry picked from commit d0f88e70f9c841bf2f8baedb89ec28523b2e357d)

qris committed:
Make BackupAccountControl throw exceptions on lock failure

Remove the compatibility code that caught these exceptions and made the
functions return 1 instead of throwing the exception. Make bbstoreaccounts
catch these exceptions and exit with status 1 instead.

qris committed:
bbstoreaccounts: more refactoring to support non-raidfile backends

qris committed:
Remove unused #includes from BackupCommands.cpp

qris committed:
Add a constant for WRITE_LOCK_FILENAME

qris committed:
Only replace getopt on Windows, not any other platform

The preprocessor test was wrong, causing this code to be built (and potentially
replace the system getopt/long) on all platforms. This made it impossible to
build with -Werror=writable-strings on Unix without modifying this file further
to fix its casting of static strings to mutable pointers.

qris committed:
Enable -Werror=writable-strings and fix one such error

qris committed:
Fix ProtocolError single-argument (subtype only) constructor

Correctly initialise mType from the ErrorType CONSTANT in the protocol message
definition in this case, instead of leaving it uninitialised. Add test. Also
test that these unexpected errors cause the connection to be closed as
expected.

Also catch SIGPIPE so that the test doesn't die when the server process closes
the connection to it.

qris committed:
Fix build order for targets which depend on auto-generated files

These auto-generated files should always be rebuilt first, and hence must be
listed first in the dependencies. Otherwise, if we add a command to a
Protocol.txt file and also implement that command in the Commands.cpp source
file, we might try to compile Commands.cpp before rebuilding
autogen_Protocol.h, which will fail with a compile error, which prevents Make
from reaching the autogen_Protocol.cpp target which would cause the header file
to be updated as a side effect, which would resolve the compile error.

qris committed:
Fix S3BackupFileSystem::GetAccountIdentifier() exception

This makes it possible to call GetAccountIdentifier() on an account which
doesn't currently have a BackupStoreInfo file, which allows us to log something
relating to the account while locking it while creating it (and the
BackupStoreInfo file).

qris committed:
Check for both SocketReadError and SocketWriteError

Depending on timing, we might get to send our message (in the test) and then
fail to read the reply, or fail during writing instead.

qris committed:
Fix GCC warnings about -Wno-* flags

GCC suppresses warnings for -Wno-* flags unless another error occurs, so we
need to use the opposite flag for detection purposes. Add an optional second
argument to BOX_CHECK_CXX_FLAG to allow specifying that detection flag.

(cherry picked from commit dd7377e3593f96b85cf82b446fe28a86f713fe61)

qris committed:
Port test_server_housekeeping to BackupFileSystem

qris committed:
Fix 5f6680520574f15b97f92e151d93efc40be0f100

Don't swallow exceptions by returning before rethrowing them

qris committed:
Port test_directory_parent_entry_tracks_directory_size to BackupFileSystem

Implement as a specialised test, and run on S3 as well.

This requires implementation of S3BackupFileSystem::DeleteFile and
DeleteDirectory, and a version of change_account_limits which takes a
reference to a BackupAccountControl.

qris committed:
Fix missing define causing unnecessary stat() on Windows

(cherry picked from commit c38e93f27ebcad6572b2cc17e4e74bf8315ddca2)

qris committed:
Remove BackupStoreAccounts::LockAccount

Replace with BackupFileSystem::GetLock where used (only in tests at this
point).

qris committed:
test/common: use setup and teardown macros on each test case

qris committed:
Add tests for exclusivity of locking

qris committed:
Mark master branch builds as full releases (not pre-releases) on GitHub

qris committed:
Remove MEMLEAKFINDER_NOT_A_LEAK from FileStream

It's not clear why we need to tell the memory leak tracking system to ignore
FileStream objects if they raise an exception in their constructors. The
objects should still be destroyed normally. Additionally, now that we
allocate more of them on the heap (for NamedLock), this can cause later
allocations to be silently ignored as not-leaks, which is causing
test_memory_leak_detection to fail.

qris committed:
Add S3 BackupFileSystem locking using SimpleDB

qris committed:
Build 64-bit OpenSSL in Windows superbuild

qris committed:
BackupStoreContext: improve log messages regarding the cache

qris committed:
BackupAccountControl: silence exceptions during CreateAccount

We expect that the raidfile or HTTP file doesn't exist during CreateAccount, so
the exception message is unhelpful, so suppress it.

qris committed:
Remove remaining calls to check_num_files() without a BackupFileSystem

qris committed:
Refactor BackupStoreInfo to remove legacy RaidFile support

qris committed:
Run test_account_limits_respected on S3 BackupFileSystem

qris committed:
Add a test for S3BackupFileSystem and its RevisionID calculation

qris committed:
Refactor testbackupstorefix to reduce code duplication

qris committed:
testbackupstorepatch: improve debugging messages

qris committed:
Port test/backupstorepatch to use BackupFileSystem

Make testbackupstorepatch do housekeeping in-process to make debugging easier.
Check return value of housekeeping to ensure that no errors were detected.

qris committed:
Improve logging in test/basicserver to help debug FreeBSD test failures

qris committed:
Move check for -Werror=narrowing to join other -Werror checks

qris committed:
Add protocol logging to help debug test/basicserver timeouts on FreeBSD

qris committed:
Remove unused #includes from BackupStoreContext.cpp

qris committed:
Add some comments

qris committed:
check_reference_counts: add assertion that ExpectedRefCounts has correct length

qris committed:
Remove full stops from ends of exception messages, for consistency

qris committed:
RaidFile: allow creating a new file with refcount == 0

Stop lying about our refcount when creating new files in BackupFileSystem.

qris committed:
Add support for general file locking in FileStream

Use this support (moved out of NamedLock) to massively simplify NamedLock.

Add support for shared file locks, and ensure that opening files respects any
exclusive (advisory) lock on that file.

qris committed:
Convert RaidFileWrite to use FileStream to write stripes

Enable exclusive locking of stripe files on all platforms, not just Windows and
O_EXLOCK.

qris committed:
Compile out test_memory_leak_detection if leak detection is disabled

qris committed:
Remove unnecessary mainhelper catch for BoxException specifically

qris committed:
Remove MemLeakFindOn/Off #includes from Utils.h

There is no code in this section, so no need to enable/disable memory leak
tracking.

qris committed:
Remove unnecessary dependencies from modules.txt

Sort dependencies into libraries first, then binaries

qris committed:
Actually pass test options (e.g. timing) to test command

qris committed:
Merge branch 'master' into s3_support_merge

qris committed:
Add more logging to test/basicserver for FreeBSD timeouts

qris committed:
Add support for S3 commands to bbstoreaccounts utility

qris committed:
Add missing #includes to BackupFileSystem.cpp

qris committed:
Remove unnecessary friend class

qris committed:
Add missing #include in BackupStoreCheck.h

qris committed:
Specialise test_multiple_uploads, but only for testbackupstore so far
